File: cielo24_.NET/WrapperCielo24/Actions.cs
using System;
using System.IO;
using System.Net;
using System.Linq;
using System.Text;
using System.Diagnostics;
using System.Security.Principal;
using System.Collections.Generic;
using WrapperCielo24.JSON;

namespace WrapperCielo24
{
    public class Actions
    {
        public const int VERSION = 1;
        private string BASE_URL = "https://api.cielo24.com";
        public string ServerUrl { get { return this.BASE_URL; } set { this.BASE_URL = value; } }
        private WebUtils web = new WebUtils();

        private const string LOGIN_PATH = "/api/account/login";
        private const string LOGOUT_PATH = "/api/account/logout";
        private const string UPDATE_PASSWORD_PATH = "/api/account/update_password";
        private const string GENERATE_API_KEY_PATH = "/api/account/generate_api_key";
        private const string REMOVE_API_KEY_PATH = "/api/account/remove_api_key";
        private const string CREATE_JOB_PATH = "/api/job/new";
        private const string AUTHORIZE_JOB_PATH = "/api/job/authorize";
        private const string DELETE_JOB_PATH = "/api/job/del";
        private const string GET_JOB_INFO_PATH = "/api/job/info";
        private const string GET_JOB_LIST_PATH = "/api/job/list";
        private const string ADD_MEDIA_TO_JOB_PATH = "/api/job/add_media";
        private const string ADD_EMBEDDED_MEDIA_TO_JOB_PATH = "/api/job/add_media_url";
        private const string GET_MEDIA_PATH = "/api/job/media";
        private const string PERFORM_TRANSCRIPTION = "/api/job/perform_transcription";
        private const string GET_TRANSCRIPTION_PATH = "/api/job/get_transcript";
        private const string GET_CAPTION_PATH = "/api/job/get_caption";
        private const string GET_ELEMENT_LIST_PATH = "/api/job/get_elementlist";
        private const string GET_LIST_OF_ELEMENT_LISTS_PATH = "/api/job/list_elementlists";

        /// ACCESS CONTROL ///

        /* Performs a Login action. If useHeaders is true, puts username and password into HTTP headers */
        public Guid Login(string username, string password, bool useHeaders = false)
        {
            this.AssertArgument(username, "Username");
            this.AssertArgument(password, "Password");

            Dictionary<string, string> queryDictionary = InitVersionDict();
            Dictionary<string, string> headers = new Dictionary<string, string>();

            if (!useHeaders)
            {
                queryDictionary.Add("username", username);
                queryDictionary.Add("password", password);
            }
            else
            {
                headers.Add("x-auth-user", username);
                headers.Add("x-auth-key", password);
            }

            Uri requestUri = Utils.BuildUri(BASE_URL, LOGIN_PATH, queryDictionary);
            string serverResponse = web.HttpRequest(requestUri, HttpMethod.GET, WebUtils.BASIC_TIMEOUT, headers);
            Dictionary<string, string> response = Utils.Deserialize<Dictionary<string, string>>(serverResponse);

            return new Guid(response["ApiToken"]);
        }

        /* Performs a Login action. If useHeaders is true, puts securekey into HTTP headers */
        public Guid Login(string username, Guid securekey, bool useHeaders = false)
        {
            this.AssertArgument(username, "Username");

            Dictionary<string, string> queryDictionary = InitVersionDict();
            Dictionary<string, string> headers = new Dictionary<string, string>();

            if (!useHeaders)
            {
                queryDictionary.Add("username", username);
                queryDictionary.Add("securekey", securekey.ToString("N"));
            }
            else
            {
                headers.Add("x-auth-user", username);
                headers.Add("x-auth-securekey", securekey.ToString("N"));
            }

            Uri requestUri = Utils.BuildUri(BASE_URL, LOGIN_PATH, queryDictionary);
            string serverResponse = web.HttpRequest(requestUri, HttpMethod.GET, WebUtils.BASIC_TIMEOUT, headers);
            Dictionary<string, string> response = Utils.Deserialize<Dictionary<string, string>>(serverResponse);

            return new Guid(response["ApiToken"]);
        }

        /* Performs a Logout action */
        public void Logout(Guid apiToken)
        {
            Dictionary<string, string> queryDictionary = this.InitAccessReqDict(apiToken);
            Uri requestUri = Utils.BuildUri(BASE_URL, LOGOUT_PATH, queryDictionary);
            web.HttpRequest(requestUri, HttpMethod.GET, WebUtils.BASIC_TIMEOUT); // Nothing returned
        }

        /* Updates password */
        public void UpdatePassword(Guid apiToken, string newPassword)
        {
            this.AssertArgument(newPassword, "New Password");

            Dictionary<string, string> queryDictionary = this.InitAccessReqDict(apiToken);
            queryDictionary.Add("new_password", newPassword);

            Uri requestUri = Utils.BuildUri(BASE_URL, UPDATE_PASSWORD_PATH, queryDictionary);
            web.HttpRequest(requestUri, HttpMethod.POST, WebUtils.BASIC_TIMEOUT, Utils.ToQuery(queryDictionary)); // Nothing returned
        }

        /* Returns a new Secure API key */
        public Guid GenerateAPIKey(Guid apiToken, string username, bool forceNew = false)
        {
            this.AssertArgument(username, "Username");

            Dictionary<string, string> queryDictionary = this.InitAccessReqDict(apiToken);
            queryDictionary.Add("account_id", username);
            queryDictionary.Add("force_new", forceNew.ToString());

            Uri requestUri = Utils.BuildUri(BASE_URL, GENERATE_API_KEY_PATH, queryDictionary);
            string serverResponse = web.HttpRequest(requestUri, HttpMethod.GET, WebUtils.BASIC_TIMEOUT);
            Dictionary<string, string> response = Utils.Deserialize<Dictionary<string, string>>(serverResponse);

            return new Guid(response["ApiKey"]);
        }

        /* Deactivates the supplied Secure API key */
        public void RemoveAPIKey(Guid apiToken, Guid apiSecurekey)
        {
            Dictionary<string, string> queryDictionary = this.InitAccessReqDict(apiToken);
            queryDictionary.Add("api_securekey", apiSecurekey.ToString("N"));

            Uri requestUri = Utils.BuildUri(BASE_URL, REMOVE_API_KEY_PATH, queryDictionary);
            web.HttpRequest(requestUri, HttpMethod.GET, WebUtils.BASIC_TIMEOUT); // Nothing returned
        }


        /// JOB CONTROL ///

        /* Creates a new job. Returns an array of Guids where 'JobId' is the 0th element and 'TaskId' is the 1st element */
        public CreateJobResult CreateJob(Guid apiToken, string jobName = null, string sourceLanguage = "en")
        {
            Dictionary<string, string> queryDictionary = this.InitAccessReqDict(apiToken);
            if (jobName != null) { queryDictionary.Add("job_name", jobName); }
            queryDictionary.Add("source_language", sourceLanguage);

            Uri requestUri = Utils.BuildUri(BASE_URL, CREATE_JOB_PATH, queryDictionary);
            string serverResponse = web.HttpRequest(requestUri, HttpMethod.GET, WebUtils.BASIC_TIMEOUT);
            CreateJobResult response = Utils.Deserialize<CreateJobResult>(serverResponse);

            return response;
        }

        /* Authorizes a job with jobId */
        public void AuthorizeJob(Guid apiToken, Guid jobId)
        {
            Dictionary<string, string> queryDictionary = InitJobReqDict(apiToken, jobId);

            Uri requestUri = Utils.BuildUri(BASE_URL, AUTHORIZE_JOB_PATH, queryDictionary);
            web.HttpRequest(requestUri, HttpMethod.GET, WebUtils.BASIC_TIMEOUT); // Nothing returned
        }

        /* Deletes a job with jobId */
        public Guid DeleteJob(Guid apiToken, Guid jobId)
        {
            Dictionary<string, string> response = getJobResponse<Dictionary<string, string>>(apiToken, jobId, DELETE_JOB_PATH);
            return new Guid(response["TaskId"]);
        }

        /* Gets information about a job with jobId */
        public JobInfo GetJobInfo(Guid apiToken, Guid jobId)
        {
            return getJobResponse<JobInfo>(apiToken, jobId, GET_ELEMENT_LIST_PATH);
        }

        /* Gets a list of jobs */
        public JobList GetJobList(Guid apiToken)
        {
            Dictionary<string, string> queryDictionary = InitAccessReqDict(apiToken);

            Uri requestUri = Utils.BuildUri(BASE_URL, GET_JOB_LIST_PATH, queryDictionary);
            string serverResponse = web.HttpRequest(requestUri, HttpMethod.GET, WebUtils.BASIC_TIMEOUT);
            JobList jobList = Utils.Deserialize<JobList>(serverResponse);

            return jobList;
        }

        /* Uploads a file from fileStream to job with jobId */
        public Guid AddMediaToJob(Guid apiToken, Guid jobId, Stream fileStream)
        {
            this.AssertArgument(fileStream, "File");

            Dictionary<string, string> queryDictionary = InitJobReqDict(apiToken, jobId);
            Uri requestUri = Utils.BuildUri(BASE_URL, ADD_MEDIA_TO_JOB_PATH, queryDictionary);
            string serverResponse = web.UploadData(requestUri, fileStream, "video/mp4");
            Dictionary<string, string> response = Utils.Deserialize<Dictionary<string, string>>(serverResponse);

            return new Guid(response["TaskId"]);
        }

        /* Provides job with jobId a url to media */
        public Guid AddMediaToJob(Guid apiToken, Guid jobId, Uri mediaUrl)
        {
            return SendMediaUrl(apiToken, jobId, mediaUrl, ADD_MEDIA_TO_JOB_PATH);
        }

        /* Provides job with jobId a url to media */
        public Guid AddEmbeddedMediaToJob(Guid apiToken, Guid jobId, Uri mediaUrl)
        {
            return SendMediaUrl(apiToken, jobId, mediaUrl, ADD_EMBEDDED_MEDIA_TO_JOB_PATH);
        }

        /* Helper method for AddMediaToJob and AddEmbeddedMediaToJob methods */
        private Guid SendMediaUrl(Guid apiToken, Guid jobId, Uri mediaUrl, string path)
        {
            this.AssertArgument(mediaUrl, "Media URL");

            Dictionary<string, string> queryDictionary = InitJobReqDict(apiToken, jobId);
            queryDictionary.Add("media_url", Utils.EncodeUrl(mediaUrl));

            Uri requestUri = Utils.BuildUri(BASE_URL, path, queryDictionary);
            string serverResponse = web.HttpRequest(requestUri, HttpMethod.GET, WebUtils.BASIC_TIMEOUT);
            Dictionary<string, string> response = Utils.Deserialize<Dictionary<string, string>>(serverResponse);

            return new Guid(response["TaskId"]);
        }

        /* Returns a Uri to the media from job with jobId */
        public Uri GetMedia(Guid apiToken, Guid jobId)
        {
            Dictionary<string, string> response = getJobResponse<Dictionary<string, string>>(apiToken, jobId, GET_MEDIA_PATH);
            return new Uri(response["MediaUrl"]);
        }

        /* Makes a PerformTranscription call */
        public Guid PerformTranscription(Guid apiToken,
                                         Guid jobId,
                                         Fidelity fidelity,
                                         Priority priority,
                                         Uri callback_uri = null,
                                         int? turnaround_hours = null,
                                         string targetLanguage = null,
                                         PerformTranscriptionOptions options = null)
        {
            Dictionary<string, string> queryDictionary = InitJobReqDict(apiToken, jobId);
            queryDictionary.Add("transcription_fidelity", fidelity.ToString());
            queryDictionary.Add("priority", priority.ToString());
            if (callback_uri != null) { queryDictionary.Add("callback_url", Utils.EncodeUrl(callback_uri)); }
            if (turnaround_hours != null) { queryDictionary.Add("turnaround_hours", turnaround_hours.ToString()); }
            if (targetLanguage != null) { queryDictionary.Add("target_language", targetLanguage); }
            if (options != null) { queryDictionary = Utils.DictConcat(queryDictionary, options.GetDictionary()); }

            Uri requestUri = Utils.BuildUri(BASE_URL, PERFORM_TRANSCRIPTION, queryDictionary);
            string serverResponse = web.HttpRequest(requestUri, HttpMethod.GET, WebUtils.BASIC_TIMEOUT);
            Dictionary<string, string> response = Utils.Deserialize<Dictionary<string, string>>(serverResponse);

            return new Guid(response["TaskId"]);
        }

        /* Returns a transcript from a job with jobId */
        public string GetTranscript(Guid apiToken, Guid jobId, TranscriptOptions transcriptOptions = null)
        {
            Dictionary<string, string> queryDictionary = InitJobReqDict(apiToken, jobId);
            if (transcriptOptions != null) { queryDictionary = Utils.DictConcat(queryDictionary, transcriptOptions.GetDictionary()); }

            Uri requestUri = Utils.BuildUri(BASE_URL, GET_TRANSCRIPTION_PATH, queryDictionary);
            return web.HttpRequest(requestUri, HttpMethod.GET, WebUtils.DOWNLOAD_TIMEOUT); // Transcript text
        }

        /* Returns a caption from a job with jobId OR if buildUri is true, returns a string representation of the uri */
        public string GetCaption(Guid apiToken, Guid jobId, CaptionFormat captionFormat, CaptionOptions captionOptions = null)
        {
            Dictionary<string, string> queryDictionary = InitJobReqDict(apiToken, jobId);
            queryDictionary.Add("caption_format", captionFormat.ToString());
            if (captionOptions != null) { queryDictionary = Utils.DictConcat(queryDictionary, captionOptions.GetDictionary()); }

            Uri requestUri = Utils.BuildUri(BASE_URL, GET_CAPTION_PATH, queryDictionary);
            string serverResponse = web.HttpRequest(requestUri, HttpMethod.GET, WebUtils.DOWNLOAD_TIMEOUT);
            if (captionOptions != null && captionOptions.BuildUrl.Equals(true))
            {
                Dictionary<string, string> response = Utils.Deserialize<Dictionary<string, string>>(serverResponse);
                return response["CaptionUrl"];
            }

            return serverResponse; // Caption text
        }

        /* Returns an element list */
        public ElementList GetElementList(Guid apiToken, Guid jobId)
        {
            return getJobResponse<ElementList>(apiToken, jobId, GET_ELEMENT_LIST_PATH);
        }

        /* Returns a list of elements lists */
        public List<ElementListVersion> GetListOfElementLists(Guid apiToken, Guid jobId)
        {
            return getJobResponse<List<ElementListVersion>>(apiToken, jobId, GET_LIST_OF_ELEMENT_LISTS_PATH);
        }


        /// PRIVATE HELPER METHODS ///

        private T getJobResponse<T>(Guid apiToken, Guid jobId, String path)
        {
            Dictionary<String, String> queryDictionary = InitJobReqDict(apiToken, jobId);
            Uri requestUri = Utils.BuildUri(BASE_URL, path, queryDictionary);
            String serverResponse = web.HttpRequest(requestUri, HttpMethod.GET, WebUtils.BASIC_TIMEOUT);
            return Utils.Deserialize<T>(serverResponse);
        }

        /* Returns a dictionary with version, api_token and job_id key-value pairs (parameters used in almost every job-control action). */
        private Dictionary<string, string> InitJobReqDict(Guid apiToken, Guid jobId)
        {
            this.AssertArgument(jobId, "Job Id");
            Dictionary<string, string> queryDictionary = this.InitAccessReqDict(apiToken);
            queryDictionary.Add("job_id", jobId.ToString("N"));
            return queryDictionary;
        }

        /* Returns a dictionary with version and api_token key-value pairs (parameters used in almost every access-control action). */
        private Dictionary<string, string> InitAccessReqDict(Guid apiToken)
        {
            this.AssertArgument(apiToken, "API Token");
            Dictionary<string, string> queryDictionary = InitVersionDict();
            queryDictionary.Add("api_token", apiToken.ToString("N"));
            return queryDictionary;
        }

        /* Returns a dictionary with version key-value pair (parameter used in every action). */
        private Dictionary<string, string> InitVersionDict()
        {
            Dictionary<string, string> queryDictionary = new Dictionary<string, string>();
            queryDictionary.Add("v", VERSION.ToString());
            return queryDictionary;
        }

        /* If arg is invalid (null or empty), throws an ArgumentException */
        private void AssertArgument(string arg, string arg_name)
        {
            if (arg == null || arg.Equals("")) { throw new ArgumentException("Invalid " + arg_name); }
        }

        private void AssertArgument(Guid arg, string arg_name)
        {
            if (arg.Equals(Guid.Empty)) { throw new ArgumentException("Invalid " + arg_name); }
        }

        private void AssertArgument(object arg, string arg_name)
        {
            if (arg == null) { throw new ArgumentException("Invalid " + arg_name); }
        }
    }
}

File: cielo24_Java/cielo24/src/cielo24/Actions.java
package cielo24;

import java.io.BufferedInputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.lang.reflect.Type;
import java.net.URL;
import java.util.ArrayList;
import java.util.HashMap;

import cielo24.WebUtils.HttpMethod;
import cielo24.json.CreateJobResult;
import cielo24.json.ElementList;
import cielo24.json.ElementListVersion;
import cielo24.json.JobInfo;
import cielo24.json.JobList;
import cielo24.options.CaptionOptions;
import cielo24.options.PerformTranscriptionOptions;
import cielo24.options.TranscriptOptions;
import cielo24.utils.Dictionary;
import cielo24.utils.Guid;
import cielo24.utils.NanoDate;
import cielo24.utils.WebException;
import cielo24.Enums.*;

public class Actions {

	private final static int API_VERSION = 1;
	public String serverUrl = "https://api.cielo24.com";
	private WebUtils web = new WebUtils();

	private final static String LOGIN_PATH = "/api/account/login";
	private final static String LOGOUT_PATH = "/api/account/logout";
	private final static String UPDATE_PASSWORD_PATH = "/api/account/update_password";
	private final static String GENERATE_API_KEY_PATH = "/api/account/generate_api_key";
	private final static String REMOVE_API_KEY_PATH = "/api/account/remove_api_key";
	private final static String CREATE_JOB_PATH = "/api/job/new";
	private final static String AUTHORIZE_JOB_PATH = "/api/job/authorize";
	private final static String DELETE_JOB_PATH = "/api/job/del";
	private final static String GET_JOB_INFO_PATH = "/api/job/info";
	private final static String GET_JOB_LIST_PATH = "/api/job/list";
	private final static String ADD_MEDIA_TO_JOB_PATH = "/api/job/add_media";
	private final static String ADD_EMBEDDED_MEDIA_TO_JOB_PATH = "/api/job/add_media_url";
	private final static String GET_MEDIA_PATH = "/api/job/media";
	private final static String PERFORM_TRANSCRIPTION = "/api/job/perform_transcription";
	private final static String GET_TRANSCRIPTION_PATH = "/api/job/get_transcript";
	private final static String GET_CAPTION_PATH = "/api/job/get_caption";
	private final static String GET_ELEMENT_LIST_PATH = "/api/job/get_elementlist";
	private final static String GET_LIST_OF_ELEMENT_LISTS_PATH = "/api/job/list_elementlists";

	public Actions() {}

	public Actions(String url) {
		this.serverUrl = url;
	}

	/// ACCESS CONTROL ///

	/* Performs a Login action. If useHeaders is true, puts username and password into HTTP headers */
	public Guid login(String username, String password, boolean useHeaders) throws IOException, WebException {
		this.assertArgument(username, "Username");
		this.assertArgument(password, "Password");

		Dictionary<String, String> queryDictionary = initVersionDict();
		Dictionary<String, String> headers = new Dictionary<String, String>();

		if (!useHeaders) {
			queryDictionary.add("username", username);
			queryDictionary.add("password", password);
		} else {
			headers.add("x-auth-user", username);
			headers.add("x-auth-key", password);
		}

		URL requestURL = Utils.buildURL(serverUrl, LOGIN_PATH, queryDictionary);
		String serverResponse = web.httpRequest(requestURL, HttpMethod.GET, WebUtils.BASIC_TIMEOUT, headers);
		HashMap<String, String> response = Utils.deserialize(serverResponse, Utils.hashMapType);

		return new Guid(response.get("ApiToken"));
	}

	/* Performs a Login action. If useHeaders is true, puts securekey into HTTP headers */
	public Guid login(String username, Guid securekey, boolean useHeaders) throws IOException, WebException {
		this.assertArgument(username, "Username");

		Dictionary<String, String> queryDictionary = initVersionDict();
		Dictionary<String, String> headers = new Dictionary<String, String>();

		if (!useHeaders) {
			queryDictionary.add("username", username);
			queryDictionary.add("securekey", securekey.toString());
		} else {
			headers.add("x-auth-user", username);
			headers.add("x-auth-securekey", securekey.toString());
		}

		URL requestURL = Utils.buildURL(serverUrl, LOGIN_PATH, queryDictionary);
		String serverResponse = web.httpRequest(requestURL, HttpMethod.GET, WebUtils.BASIC_TIMEOUT, headers);
		HashMap<String, String> response = Utils.deserialize(serverResponse, Utils.hashMapType);

		return new Guid(response.get("ApiToken"));
	}

	/* Performs a Logout action */
	public void logout(Guid apiToken) throws IOException, WebException {
		Dictionary<String, String> queryDictionary = this.initAccessReqDict(apiToken);
		URL requestURL = Utils.buildURL(serverUrl, LOGOUT_PATH, queryDictionary);
		web.httpRequest(requestURL, HttpMethod.GET, WebUtils.BASIC_TIMEOUT); // Nothing returned
	}

	/* Updates password */
	public void updatePassword(Guid apiToken, String newPassword) throws IOException, WebException {
		this.assertArgument(newPassword, "New Password");

		Dictionary<String, String> queryDictionary = this.initAccessReqDict(apiToken);
		queryDictionary.add("new_password", newPassword);

		URL requestURL = Utils.buildURL(serverUrl, UPDATE_PASSWORD_PATH, queryDictionary);
		web.httpRequest(requestURL, HttpMethod.POST, WebUtils.BASIC_TIMEOUT, Utils.toQuery(queryDictionary)); // Nothing returned
	}

	/* Returns a new Secure API key */
	public Guid generateAPIKey(Guid apiToken, String username, boolean forceNew) throws IOException, WebException {
		this.assertArgument(username, "Username");

		Dictionary<String, String> queryDictionary = this.initAccessReqDict(apiToken);
		queryDictionary.add("account_id", username);
		queryDictionary.add("force_new", Boolean.toString(forceNew));

		URL requestURL = Utils.buildURL(serverUrl, GENERATE_API_KEY_PATH, queryDictionary);
		String serverResponse = web.httpRequest(requestURL, HttpMethod.GET, WebUtils.BASIC_TIMEOUT);
		HashMap<String, String> response = Utils.deserialize(serverResponse, Utils.hashMapType);

		return new Guid(response.get("ApiKey"));
	}

	/* Deactivates the supplied Secure API key */
	public void removeAPIKey(Guid apiToken, Guid apiSecurekey) throws IOException, WebException {
		Dictionary<String, String> queryDictionary = this.initAccessReqDict(apiToken);
		queryDictionary.add("api_securekey", apiSecurekey.toString());

		URL requestURL = Utils.buildURL(serverUrl, REMOVE_API_KEY_PATH, queryDictionary);
		web.httpRequest(requestURL, HttpMethod.GET, WebUtils.BASIC_TIMEOUT); // Nothing returned
	}

	/// JOB CONTROL ///

	/* Creates a new job. Returns an array of Guids where 'JobId' is the 0th element and 'TaskId' is the 1st element */
	public CreateJobResult createJob(Guid apiToken, String jobName, String language) throws IOException, WebException {
		Dictionary<String, String> queryDictionary = this.initAccessReqDict(apiToken);
		if (jobName != null) {
			queryDictionary.add("job_name", jobName);
		}
		queryDictionary.add("language", language);

		URL requestURL = Utils.buildURL(serverUrl, CREATE_JOB_PATH, queryDictionary);
		String serverResponse = web.httpRequest(requestURL, HttpMethod.GET, WebUtils.BASIC_TIMEOUT);
		CreateJobResult response = Utils.deserialize(serverResponse, CreateJobResult.class);

		return response;
	}

	/* Authorizes a job with jobId */
	public void authorizeJob(Guid apiToken, Guid jobId) throws IOException, WebException {
		Dictionary<String, String> queryDictionary = initJobReqDict(apiToken, jobId);
		URL requestURL = Utils.buildURL(serverUrl, AUTHORIZE_JOB_PATH, queryDictionary);
		web.httpRequest(requestURL, HttpMethod.GET, WebUtils.BASIC_TIMEOUT); // Nothing returned
	}

	/* Deletes a job with jobId */
	public Guid deleteJob(Guid apiToken, Guid jobId) throws IOException, WebException {
		HashMap<String, String> response = getJobResponse(apiToken, jobId, DELETE_JOB_PATH, Utils.hashMapType);
		return new Guid(response.get("TaskId"));
	}

	/* Gets information about a job with jobId */
	public JobInfo getJobInfo(Guid apiToken, Guid jobId) throws IOException, WebException {
		return getJobResponse(apiToken, jobId, GET_JOB_INFO_PATH, JobInfo.class);
	}

	/* Gets a list of jobs */
	public JobList getJobList(Guid apiToken) throws IOException, WebException {
		Dictionary<String, String> queryDictionary = initAccessReqDict(apiToken);
		URL requestURL = Utils.buildURL(serverUrl, GET_JOB_LIST_PATH, queryDictionary);
		String serverResponse = web.httpRequest(requestURL, HttpMethod.GET, WebUtils.BASIC_TIMEOUT);
		JobList jobList = Utils.deserialize(serverResponse, JobList.class);
		return jobList;
	}

	/* Uploads a file from fileStream to job with jobId */
	public Guid addMediaToJob(Guid apiToken, Guid jobId, File file) throws IOException, WebException {
		this.assertArgument(file, "Local Media File");
		BufferedInputStream stream = new BufferedInputStream(new FileInputStream(file));

		Dictionary<String, String> queryDictionary = initJobReqDict(apiToken, jobId);
		URL requestURL = Utils.buildURL(serverUrl, ADD_MEDIA_TO_JOB_PATH, queryDictionary);
		String serverResponse = web.uploadData(requestURL, stream, "video/mp4", file.length());
		HashMap<String, String> response = Utils.deserialize(serverResponse, Utils.hashMapType);

		return new Guid(response.get("TaskId"));
	}

	/* Provides job with jobId a url to media */
	public Guid addMediaToJob(Guid apiToken, Guid jobId, URL mediaUrl) throws IOException, WebException {
		return sendMediaUrl(apiToken, jobId, mediaUrl, ADD_MEDIA_TO_JOB_PATH);
	}

	/* Provides job with jobId a url to media */
	public Guid addEmbeddedMediaToJob(Guid apiToken, Guid jobId, URL mediaUrl) throws IOException, WebException {
		return sendMediaUrl(apiToken, jobId, mediaUrl, ADD_EMBEDDED_MEDIA_TO_JOB_PATH);
	}

	/* Helper method for AddMediaToJob and AddEmbeddedMediaToJob methods */
	private Guid sendMediaUrl(Guid apiToken, Guid jobId, URL mediaUrl, String path) throws IOException, WebException {
		this.assertArgument(mediaUrl, "Media URL");

		Dictionary<String, String> queryDictionary = initJobReqDict(apiToken, jobId);
		queryDictionary.add("media_url", Utils.encodeUrl(mediaUrl));

		URL requestURL = Utils.buildURL(serverUrl, path, queryDictionary);
		String serverResponse = web.httpRequest(requestURL, HttpMethod.GET, WebUtils.BASIC_TIMEOUT);
		HashMap<String, String> response = Utils.deserialize(serverResponse, Utils.hashMapType);

		return new Guid(response.get("TaskId"));
	}

	/* Returns a URL to the media from job with jobId */
	public URL getMedia(Guid apiToken, Guid jobId) throws IOException, WebException {
		HashMap<String, String> response = getJobResponse(apiToken, jobId, GET_MEDIA_PATH, Utils.hashMapType);
		return new URL(response.get("MediaUrl"));
	}

	/* Makes a PerformTranscription call */
	public Guid performTranscription(Guid apiToken,
									 Guid jobId,
									 Fidelity fidelity,
									 Priority priority,
									 URL callback_uri,
									 Integer turnaround_hours,
									 String targetLanguage,
									 PerformTranscriptionOptions options)
									 throws IOException, WebException, IllegalArgumentException, IllegalAccessException {
		Dictionary<String, String> queryDictionary = initJobReqDict(apiToken, jobId);
		queryDictionary.add("transcription_fidelity", fidelity.toString());
		queryDictionary.add("priority", priority.toString());
		if (callback_uri != null) {
			queryDictionary.add("callback_url", Utils.encodeUrl(callback_uri));
		}
		if (turnaround_hours != null) {
			queryDictionary.add("turnaround_hours", turnaround_hours.toString());
		}
		if (targetLanguage != null) {
			queryDictionary.add("target_language", targetLanguage);
		}
		if (options != null) {
			queryDictionary.addAll(options.GetDictionary());
		}

		URL requestURL = Utils.buildURL(serverUrl, PERFORM_TRANSCRIPTION, queryDictionary);
		String serverResponse = web.httpRequest(requestURL, HttpMethod.GET, WebUtils.BASIC_TIMEOUT);
		HashMap<String, String> response = Utils.deserialize(serverResponse, Utils.hashMapType);

		return new Guid(response.get("TaskId"));
	}

	/* Returns a transcript from a job with jobId */
	public String getTranscript(Guid apiToken,
								Guid jobId,
								TranscriptOptions transcriptOptions)
								throws IOException, WebException, IllegalArgumentException, IllegalAccessException {
		Dictionary<String, String> queryDictionary = initJobReqDict(apiToken, jobId);
		if (transcriptOptions != null) {
			queryDictionary.addAll(transcriptOptions.GetDictionary());
		}

		URL requestURL = Utils.buildURL(serverUrl, GET_TRANSCRIPTION_PATH, queryDictionary);
		return web.httpRequest(requestURL, HttpMethod.GET, WebUtils.DOWNLOAD_TIMEOUT); // Transcript text
	}

	/* Returns a caption from a job with jobId OR if buildURL is true, returns a String representation of the URL */
	public String getCaption(Guid apiToken,
							 Guid jobId,
							 CaptionFormat captionFormat,
							 CaptionOptions captionOptions)
							 throws IOException, WebException, IllegalArgumentException, IllegalAccessException {
		Dictionary<String, String> queryDictionary = initJobReqDict(apiToken, jobId);
		queryDictionary.add("caption_format", captionFormat.toString());
		if (captionOptions != null) {
			queryDictionary.addAll(captionOptions.GetDictionary());
		}

		URL requestURL = Utils.buildURL(serverUrl, GET_CAPTION_PATH, queryDictionary);
		String serverResponse = web.httpRequest(requestURL, HttpMethod.GET, WebUtils.DOWNLOAD_TIMEOUT);
		if (captionOptions != null && captionOptions.buildUrl != null && captionOptions.buildUrl) {
			HashMap<String, String> response = Utils.deserialize(serverResponse, Utils.hashMapType);
			return response.get("CaptionUrl");
		}

		return serverResponse; // Caption text
	}

	/* Returns an element list */
	public ElementList getElementList(Guid apiToken, Guid jobId, NanoDate elementListVersion) throws IOException, WebException {
		Dictionary<String, String> queryDictionary = initJobReqDict(apiToken, jobId);
		if (elementListVersion != null) {
			queryDictionary.add("elementlist_version", elementListVersion.toString());
		}

		URL requestURL = Utils.buildURL(serverUrl, GET_ELEMENT_LIST_PATH, queryDictionary);
		String serverResponse = web.httpRequest(requestURL, HttpMethod.GET, WebUtils.BASIC_TIMEOUT);
		return Utils.deserialize(serverResponse, ElementList.class);
	}

	/* Returns a list of elements lists */
	public ArrayList<ElementListVersion> getListOfElementLists(Guid apiToken, Guid jobId) throws IOException, WebException {
		return getJobResponse(apiToken, jobId, GET_LIST_OF_ELEMENT_LISTS_PATH, Utils.listELType);
	}

	/// OVERLOADED VERISONS ///
	/////////////////////////////////////////////////////////////////////
	public Guid login(String username, String password) throws IOException, WebException {
		return login(username, password, false);
	}

	public Guid login(String username, Guid securekey) throws IOException, WebException {
		return login(username, securekey, false);
	}

	public Guid generateAPIKey(Guid apiToken, String username) throws IOException, WebException {
		return generateAPIKey(apiToken, username, false);
	}

	public CreateJobResult createJob(Guid apiToken) throws IOException, WebException {
		return createJob(apiToken, null, "en");
	}

	public String getTranscript(Guid apiToken, Guid jobId) throws IOException, WebException, IllegalArgumentException, IllegalAccessException {
		return getTranscript(apiToken, jobId, null);
	}

	public String getCaption(Guid apiToken, Guid jobId, CaptionFormat captionFormat)
							 throws IOException, WebException, IllegalArgumentException, IllegalAccessException {
		return getCaption(apiToken, jobId, captionFormat, null);
	}

	public ElementList getElementList(Guid apiToken, Guid jobId) throws IOException, WebException {
		return getElementList(apiToken, jobId, null);
	}

	public Guid performTranscription(Guid apiToken, Guid jobId, Fidelity fidelity, Priority priority)
									 throws IOException, WebException, IllegalArgumentException, IllegalAccessException {
		return performTranscription(apiToken, jobId, fidelity, priority, null, null, null, null);
	}

	/////////////////////////////////////////////////////////////////////

	/// PRIVATE HELPER METHODS ///

	private <T> T getJobResponse(Guid apiToken, Guid jobId, String path, Type type) throws IOException, WebException {
		Dictionary<String, String> queryDictionary = initJobReqDict(apiToken, jobId);
		URL requestURL = Utils.buildURL(serverUrl, path, queryDictionary);
		String serverResponse = web.httpRequest(requestURL, HttpMethod.GET, WebUtils.BASIC_TIMEOUT);
		return Utils.deserialize(serverResponse, type);
	}

	/* Returns a dictionary with version, api_token and job_id key-value pairs (parameters used in almost every job-control action). */
	private Dictionary<String, String> initJobReqDict(Guid apiToken, Guid jobId) {
		this.assertArgument(jobId, "Job Id");
		Dictionary<String, String> queryDictionary = this.initAccessReqDict(apiToken);
		queryDictionary.add("job_id", jobId.toString());
		return queryDictionary;
	}

	/* Returns a dictionary with version and api_token key-value pairs (parameters used in almost every access-control action). */
	private Dictionary<String, String> initAccessReqDict(Guid apiToken) {
		this.assertArgument(apiToken, "API Token");
		Dictionary<String, String> queryDictionary = initVersionDict();
		queryDictionary.add("api_token", apiToken.toString());
		return queryDictionary;
	}

	/* Returns a dictionary with version key-value pair (parameter used in every action). */
	private Dictionary<String, String> initVersionDict() {
		Dictionary<String, String> queryDictionary = new Dictionary<String, String>();
		queryDictionary.add("v", Integer.toString(API_VERSION));
		return queryDictionary;
	}

	/* If arg is invalid (null or empty), throws an IllegalArgumentException */
	private void assertArgument(String arg, String arg_name) {
		if (arg == null || arg.equals("")) {
			throw new IllegalArgumentException("Invalid " + arg_name);
		}
	}

	private void assertArgument(Object arg, String arg_name) {
		if (arg == null) {
			throw new IllegalArgumentException("Invalid " + arg_name);
		}
	}
}